pyclibrary/c_model.py

```diff
@@ -0,0 +1,589 @@
+    "C" type.
+    """
+
+    __slots__ = ['quals']
+
+    def __init__(self, quals=None):
+        """
+        :param list[str]|None quals: a list of type qualifiers
+            attached to this type.
+        """
+        self.quals = quals or []
+
+    def resolve(self, typedefs):
+        """Resolves all references to type definitions. A reference to
+        a type definition is represented by the TypeRef object. All
+        CustomType objects are resolved with the help of typedefs.
+
+        :param dict[str, CLibType] typedefs: maps the name of all known
+            to the corresponding CLibType.
+        :returns: a CLibType descendant object, that contains no TypeRef's.
+        :rtype: CLibType
+        """
+        return self
+
+    def c_repr(self, referrer_c_repr=None):
+        """Get the c representation of this type.
+
+        :param str|None referrer_c_repr: Has to be either the
+            referrer_c_repr part of the type definition or None,
+            if this is an abstract type:
+            * referrer_c_repr='a' => "int *a[2]"
+            * referrer_c_repr=None => "int *[2]"
+        :rtype: str
+        """
+        raise NotImplementedError()
+
+    def __str__(self):
+        return self.c_repr()
+
+
+class SimpleType(CLibType):
+    """Abstract Base class for non-composed types
+    (BuildinType and CustomType)
+    """
+
+    __slots__ = ['type_name']
+
+    def __init__(self, type_name, quals=None):
+        super(SimpleType, self).__init__(quals)
+        self.type_name = type_name
+
+    def c_repr(self, referrer_c_repr=None):
+        return (' '.join(self.quals + [self.type_name]) +
+                _lpadded_str(referrer_c_repr))
+
+
+class BuiltinType(SimpleType):
+    """Is used for modeling scalar C types like:"
+     * "``[signed/unsigned] char``"
+     * "``[signed/unsigned] short [int]``"
+     * "``[signed/unsigned] int``"
+     * "``[signed/unsigned] long [int]``"
+     * "``float``"
+     * "``double``"
+     * ...
+
+    The class does no checks if the *type_name* is actually a valid C
+    scalar type.
+    """
+    pass
+
+
+class CustomType(SimpleType):
+    """A reference to a custom type by name.
+
+     These references have to follow one of the following forms:
+
+     * "``<typedefname>``"
+     * "``struct <structname>``"
+     * "``union <unionname>``"
+     * "``enum <enumname>``"
+    """
+
+    def resolve(self, typedefs):
+        if self.type_name not in typedefs:
+            raise UnknownCustomType('{!r} is a unknown type'
+                                    .format(self.type_name))
+
+        result = typedefs[self.type_name].resolve(typedefs)
+
+        if self.quals:
+            # do not modify result directly to avoid sideeffects
+            cloned_result = result.copy()
+
+            cloned_result.quals = self.quals + result.quals
+            return cloned_result
+        else:
+            return result
+
+
+# noinspection PyAbstractClass
```

Comment:
Sorry, its a PyCharm proprietary code annotation (I am used to insert them, since in our company we all use PyCharm)
I will remove it...